### PR TITLE
plugin/kubernetes: Modify integration tests for coredns/ci

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. What does this pull request do?

Modifications to kubernetes integration tests for integration with coredns/ci.
This is paired with PR coredns/ci#2

### 2. Which issues (if any) are related?
#902

### 3. Which documentation changes (if any) need to be made?

nothing user facing.  may need to update dev docs

### CODE REVIEW

File: test/kubernetes/a_test.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
+// +build ci
+
+package kubernetes
+
+import (
+	"testing"
+
+	"github.com/coredns/coredns/plugin/test"
+
+	testk8s "github.com/coredns/ci/test/kubernetes"
```
Comment (miekg) [inline test/kubernetes/a_test.go]: fmt

File: test/kubernetes/a_test.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
+// +build ci
```
Comment (miekg) [inline test/kubernetes/a_test.go]: address_test.go would be more descriptive.

File: test/kubernetes/a_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            upstream ` + udp + `
+        }
+    }
+`
+
+	err := testk8s.LoadCorefile(corefile)
+	if err != nil {
+		t.Fatalf("Could not load corefile: %s", err)
+	}
+	testk8s.DoIntegrationTests(t, dnsTestCasesA, "test-1")
```
Comment (miekg) [inline test/kubernetes/a_test.go]: Can we please stop hidding away the test and just put the code here. I'm assuming that this  (DoIntegrationTests) does all the sorting, including any CNAME sorting by default.

Just copy those few lines here.

File: test/kubernetes/a_test.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        errors
+        log
+        kubernetes cluster.local 10.in-addr.arpa {
+            namespaces test-1
+            upstream ` + udp + `
+        }
+    }
+`
+
+	err := testk8s.LoadCorefile(corefile)
```
Comment (miekg) [inline test/kubernetes/a_test.go]: why can't we use the functions we have in test ?

**Review by miekg — CHANGES_REQUESTED** [review]

File: test/kubernetes/a_test.go (hunk 4 above)
Comment (chrisohaver) [reply test/kubernetes/a_test.go]: These are the same functions, moved to the ci repo, as you had requested.

File: test/kubernetes/a_test.go (hunk 2 above)
Comment (chrisohaver) [reply test/kubernetes/a_test.go]: the A is for type "A" .... but I can change it to "address"

File: test/kubernetes/a_test.go (hunk 1 above)
Comment (chrisohaver) [reply test/kubernetes/a_test.go]: can you clarify the comment?

File: test/kubernetes/a_test.go (hunk 3 above)
Comment (chrisohaver) [reply test/kubernetes/a_test.go]: `DoIntegrationTests` is ~50  lines. It does the following...

- Starts the dns client pod
- The for each test case: 
  - Builds/executes dig command 
  - parses response, sorts, and compares to expected

File: test/kubernetes/a_test.go (hunk 4 above)
Comment (chrisohaver) [reply test/kubernetes/a_test.go]: To clarify: These are "test framework" functions, (like dig parsing), which are in the ci repo.

File: test/kubernetes/a_test.go (hunk 1 above)
Comment (chrisohaver) [reply test/kubernetes/a_test.go]: Do you mean that the `github.com/coredns/ci/test/kubernetes` import should be grouped with the `github.com/coredns/coredns/plugin/test` import?

File: test/kubernetes/a_test.go (hunk 3 above)
Comment (chrisohaver) [reply test/kubernetes/a_test.go]: I can reorganize things so that the tests themselves are executed in the test files...

Comment (chrisohaver) [reply test/kubernetes/a_test.go]: To be more precise, you want that all test failures and errors are generated directly from the X_test.go files?

File: Makefile [hunk 5]
```diff
@@ -54,6 +54,9 @@ ifeq ($(TEST_TYPE),coverage)
 	done
 endif
 
+.PHONY: ci
```
Comment (miekg) [inline Makefile]: Why can't we re-use the existing tag?

I'm sorry, I completely fail to see why (at this stage) we need changes *at all* to coredns' source

Comment (johnbelamaric) [reply Makefile]: Chris is trying to get the K8s integration testing out of the main repo and out of the main Travis build, which is what you asked for. I think he would have been happy to leave it all in here, but is trying to make it work in the way we understood you wanted. Maybe you can provide more detail on exactly what you are looking for, so we can stop going back and forth on this and wasting time?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/kubern..." ]
>+.PHONY: ci
>
>Chris is trying to get the K8s integration testing out of the main repo and out of the main Travis build, which is what you asked for. I think he would have been happy to leave it all in here, but is trying to make it work in the way we understood you wanted. Maybe you can provide more detail on exactly what you are looking for, so we can stop going back and forth on this and wasting time?

Ack. So proposal:

1. copy the few k8s test to the ci repo
2. get the kinks worked out there
3. remove from coredns/coredns
4. move etcd test as well

I'm wondering how white boxy the intergration tests will be anyway. I.e. how far 
off will those be from starting CoreDNS with a corefile and replying queries? 
That would imply you *only* need the latest coredns binary (possibly as a 
docker container).

Open question is where should reply data live, i.e in coredns/ci or 
coredns/coredns?

**Comment by johnbelamaric** [comment]
> That would imply you *only* need the latest coredns binary (possibly as a
docker container).

We want to run these against PRs. You comment on the PR /integration and it will run them (@chrisohaver is that right or is it automatic?)

**Comment by chrisohaver** [comment]
> We want to run these against PRs. You comment on the PR /integration and it will run them (@chrisohaver is that right or is it automatic?)

That is correct.

**Pull request merged**